### PR TITLE
Search with special characters gives bad query

### LINKED ISSUE #863: Search with special characters gives bad query
Poor title (again, sorry). Do we need an error message like the 🐣 ❓  or just allow special characters?
You can search with a **+** and **&** on _snapcraft.io_ although the results are the same as when you search without.

image

### PR DESCRIPTION
**Pull request by Feichtmeier**
image
image
image

Fixes #863

### CODE REVIEW

File: lib/app/explore/explore_model.dart [hunk 1]
```diff
@@ -128,9 +128,11 @@ class ExploreModel extends SafeChangeNotifier {
     if (searchQuery.isEmpty) {
       return [];
     } else {
+      final clearedSearchQuery =
+          searchQuery.replaceAll(RegExp('[^A-Za-z0-9]', unicode: true), '');
```
Comment (d-loose) [inline lib/app/explore/explore_model.dart:132]: We also need to preserve whitespace, otherwise entering 'foo bar' will only give you results for 'foobar'

**Review by d-loose — CHANGES_REQUESTED** [review]
Should we apply the same search string cleaning to `findAppstreamComponents` for the sake of consistency?

File: lib/app/explore/explore_model.dart (hunk 1 above)
Comment (Feichtmeier) [reply lib/app/explore/explore_model.dart:132]: whats the regex for this?

**Comment by Feichtmeier** [comment]
🤔 thinking about this again I think we shouldn't do this here at all. IMO this is something for snapd or snapd.dart
We just send strings

File: lib/app/explore/explore_model.dart (hunk 1 above)
Comment (d-loose) [reply lib/app/explore/explore_model.dart:132]: I'd propose

```dart
replaceAll(RegExp(r'[^\w\s]+'), ' ', unicode: true)
```
to replace all non-alphanumeric and non-whitespace characters by a single space.

Comment (Feichtmeier) [reply lib/app/explore/explore_model.dart:132]: Problem is for example xournal++

**Comment by Feichtmeier** [comment]
Can I close this @d-loose ? I think this is the wrong approach and the wrong repository

**Comment by d-loose** [comment]
I'm trying to figure out where the problem is - dart's `Uri` class should take care of correctly encoding the query parameters when sending a request to the snapd API..
I'd leave it open until we have a fix (here or elsewhere) - the user definitely shouldn't see a 'Bad request' instead of search results.

**Pull request closed without merging**